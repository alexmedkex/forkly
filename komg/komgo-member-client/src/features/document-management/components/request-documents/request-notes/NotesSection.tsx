import * as React from 'react'
import { Form, TextArea } from 'semantic-ui-react'
import styled from 'styled-components'

import { SectionCard } from '../SectionCard'
import { IncomingNotes } from './IncomingNotes'
import { NoteIndexControl } from './NoteIndexControl'
import { Note } from '../../../store'

export interface Props {
  noteInput: Note | null
  notes: Note[]
  setNoteContent(content: string): void
  getCounterpartyNameById(counterpartyId: string): string
}

export const NotesSection = (props: Props) => {
  const notes = [...props.notes] || []
  const { noteInput } = props

  const [textInput, setInput] = React.useState(noteInput ? noteInput.content : '')

  const hasNotes = notes && notes.length > 0
  const [noteIndex, setIndex] = React.useState(0)

  const atLatestNote = hasNotes && noteIndex + 1 === notes.length

  /*
    Track whether the textbox has focus for case where
    - replying to last existing note
    - textbox should expand from one row to three, shrinking area above
  */
  const [textAreaHasfocus, setFocus] = React.useState(false)

  return (
    <SectionCard title={renderNotesSectionTitle(noteIndex, setIndex, notes.length)}>
      <Form data-test-id="request-notes-form">
        {hasNotes && (
          <IncomingNotes
            note={notes[noteIndex || 0]}
            noteIndex={noteIndex}
            getCounterpartyNameById={props.getCounterpartyNameById}
            textAreaHasFocus={textAreaHasfocus}
            atLatestNote={atLatestNote}
          />
        )}
        {hasNotes ? (
          atLatestNote && (
            <ResizeTextArea
              data-test-id="request-notes-input"
              className="style-scroll"
              placeholder={REPLY_HERE}
              value={textInput}
              onInput={e => setInput(e.target.value)}
              onFocus={() => setFocus(true)}
              onBlur={() => {
                setFocus(false)
                props.setNoteContent(textInput)
              }}
              rows={12}
            />
          )
        ) : (
          <NoResizeTextArea
            data-test-id="request-notes-input"
            className="style-scroll"
            placeholder={TYPE_HERE}
            value={textInput}
            onInput={e => setInput(e.target.value)}
            onBlur={() => props.setNoteContent(textInput)}
            rows={13}
          />
        )}
      </Form>
    </SectionCard>
  )
}

const renderNotesSectionTitle = (noteIndex: number, setIndex: (n: number) => void, maxIndex: number) => {
  return (
    <Title>
      <p>NOTES</p>
      {maxIndex > 0 && <NoteIndexControl currentIndex={noteIndex} maxIndex={maxIndex} setIndex={setIndex} />}
    </Title>
  )
}

const TYPE_HERE = 'Type here...'

const REPLY_HERE = 'Reply here...'

/**
 * Yes, we actually need two TextArea components.
 * One which is always the same size
 * One which resizes when it receives focus
 * But not from the original size, oh no, from a new one.
 */
const NoResizeTextArea = styled(TextArea)`
  &&& {
    resize: none;
    overflow-y: auto;
  }
`

const ResizeTextArea = styled(TextArea)`
  &&& {
    resize: none;
    height: 3em;
    overflow-y: auto;

    &:focus {
      height: 7em;
    }
  }
`

const Title = styled.div`
  display: flex;
  justify-content: space-between;
  width: 105%;
`
