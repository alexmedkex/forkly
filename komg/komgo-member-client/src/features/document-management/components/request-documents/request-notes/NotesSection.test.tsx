import * as React from 'react'
import * as renderer from 'react-test-renderer'

import { NotesSection, Props } from './NotesSection'

describe('NotesSection', () => {
  it('should match default snapshot', () => {
    const props: Props = {
      notes: [],
      noteInput: null,
      getCounterpartyNameById: jest.fn(str => 'whatever'),
      setNoteContent: jest.fn((content: string) => void 0)
    }
    expect(renderer.create(<NotesSection {...props} />).toJSON()).toMatchSnapshot()
  })
})
