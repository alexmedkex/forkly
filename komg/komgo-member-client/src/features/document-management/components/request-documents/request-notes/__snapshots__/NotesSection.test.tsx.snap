// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotesSection should match default snapshot 1`] = `
.c1.c1.c1.c1.c1.c1.c1.c1.c1.c1.c1 {
  margin: 0px;
  padding: 16px 30px 30px 30px;
}

.c0.c0.c0.c0.c0.c0.c0.c0.c0.c0.c0 {
  height: 325px;
  width: unset;
  margin: 30px;
  padding-left: 0px;
  padding-rigth: 0px;
  -webkit-box-shadow: 0px 1px 11px -4px rgba(0,0,0,0.75);
  -moz-box-shadow: 0px 1px 11px -4px rgba(0,0,0,0.75);
  box-shadow: 0px 1px 11px -4px rgba(0,0,0,0.75);
}

.c2.c2.c2.c2.c2.c2.c2.c2.c2.c2.c2 {
  font-size: 11px;
  height: 21px;
  width: 549px;
  color: #1c2936;
  font-weight: 600;
  line-height: 21px;
}

.c4.c4.c4 {
  resize: none;
  overflow-y: auto;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 105%;
}

<div
  className="ui card c0"
  onClick={[Function]}
  style={Object {}}
>
  <div
    className="content c1"
  >
    <div
      className="header c2"
    >
      <div
        className="c3"
      >
        <p>
          NOTES
        </p>
      </div>
    </div>
    <div
      className="description "
    >
      <form
        className="ui form"
        data-test-id="request-notes-form"
        onSubmit={[Function]}
      >
        <textarea
          className="style-scroll c4"
          data-test-id="request-notes-input"
          onBlur={[Function]}
          onChange={[Function]}
          onInput={[Function]}
          placeholder="Type here..."
          rows={13}
          style={
            Object {
              "resize": "",
            }
          }
          value=""
        />
      </form>
    </div>
  </div>
</div>
`;
